lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,427 @@
+        if ((param->ulWidthInBits % 8) != 0) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+
+        /* Note that section 5.1 denotes the maximum length of the counter
+         * to be 32. */
+        if (param->ulWidthInBits > 32) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+        break;
+                                        }
+    case CK_SP800_108_DKM_LENGTH: {
+        if (data->ulValueLen != sizeof(CK_SP800_108_DKM_LENGTH_FORMAT)) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+
+        CK_SP800_108_DKM_LENGTH_FORMAT_PTR param = (CK_SP800_108_DKM_LENGTH_FORMAT_PTR) data->pValue;
+
+        /* Validate the method parameter. */
+        if (param->dkmLengthMethod != CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS &&
+            param->dkmLengthMethod != CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS
+        ) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+
+        /* Validate the endian parameter. */
+        if (param->bLittleEndian != CK_TRUE && param->bLittleEndian != CK_FALSE) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+
+        /* Validate the maximum width: we restrict it to being a byte width
+         * instead of a bit width due to restrictions by the underlying
+         * PRFs. */
+        if ((param->ulWidthInBits % 8) != 0) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+
+        /* Ensure that the width doesn't overflow a 64-bit int. This
+         * restriction is arbitrary but since the counters can't exceed
+         * 32-bits (and most PRFs output at most 1024 bits), you're unlikely
+         * to need all 64-bits of length indicator. */
+        if (param->ulWidthInBits > 64) {
+            return CKR_MECHANISM_PARAM_INVALID;
+        }
+        break;
+                                  }
+    case CK_SP800_108_BYTE_ARRAY:
+        /* There is no additional data to process here. */
+        break;
+    }
+
+    return CKR_OK;
+}
+
+static CK_RV derived_validate(CK_DERIVED_KEY_PTR key) {
+    return CKR_FUNCTION_NOT_SUPPORTED;
+}
+
+static CK_RV counter_validate(CK_SP800_108_KDF_PARAMS_PTR params) {
+    CK_RV ret = CKR_MECHANISM_PARAM_INVALID;
+    int param_type_count[5]  = {0, 0, 0, 0, 0};
+    size_t offset = 0;
+
+    /* Start with checking the prfType as a mechanism against a list of
+     * PRFs allowed by PKCS#11 v3.0. */
+    if (params->prfType != CKM_SHA_1_HMAC &&
+        params->prfType != CKM_SHA224_HMAC &&
+        params->prfType != CKM_SHA256_HMAC &&
+        params->prfType != CKM_SHA384_HMAC &&
+        params->prfType != CKM_SHA512_HMAC &&
+        /* The following types aren't defined in NSS yet. */
+        /* params->prfType != CKM_SHA3_224_HMAC && */
+        /* params->prfType != CKM_SHA3_256_HMAC && */
+        /* params->prfType != CKM_SHA3_384_HMAC && */
+        /* params->prfType != CKM_SHA3_512_HMAC && */
+        /* params->prfType != CKM_3DES_CMAC && */
+        params->prfType != CKM_AES_CMAC
```

Comment:
Good point. Where does "glue" code such as this live? Should I just put it in `lib/softoken/pkcs11c.c`?

The implementation in `sftkike.c` doesn't seem to validate if something is a valid (H)MAC and doesn't currently support CMAC (but does support AES XCBC). `sftk_map_hmac_to_hash` could be reused for the latter HMAC->Hash mapping. There's a comment there that says: 

```c
/* Map the mechanism to the underlying hash. If the type is not a hash
 * or HMAC, return HASH_AlgNULL. This can happen legitimately if
 * we are doing AES XCBC */
```

So we could probably extend it and turn it into a proper mapper / validator, at least for HMACs.

Reply:
yes, I think this can move to probably pkcs11u.c and add it to the pkcs11i.h header. A mapping of hmac to Hash_alg mechanism is generally useful, Something that gets a general mac interface is better. Even with the map_hmac_to_hash function, just checking that the mechansim is an hmac, then handling the cmac separately is a win.

Reply:
How sftk_ike handles the parameter tests:
static CK_RV
prf_setup(prfContext *context, CK_MECHANISM_TYPE mech)
{
    context->hashType = sftk_map_hmac_to_hash(mech);
    context->hashObj = NULL;
    context->hmac = NULL;
    context->aes = NULL;
    if (context->hashType != HASH_AlgNULL) {
        context->hashObj = HASH_GetRawHashObject(context->hashType);
        if (context->hashObj == NULL) {
            return CKR_GENERAL_ERROR;
        }
        return CKR_OK;
    } else if (mech == CKM_AES_XCBC_MAC) {
        return CKR_OK;
    }
    return CKR_MECHANISM_PARAM_INVALID;
}

Reply:
We could take the prf_ functions in ike and move them to pkcs11i.h and then add cmac to them. That addes cmac to ike and aes_xcbc to sp100-108.

Reply:
I don't personally think there's much value in extending SP800-108 to use AES-XCBC as a mechanism type. We certainly could, and that's PKCS#11's prerogative. But XCBC (CBC-MAC / OMAC1) isn't FIPS standardized (whether that's good or bad is subjective). 

Regardless, its use seems limited to [IPSec](https://tools.ietf.org/html/rfc3566) and other major protocols like TLS seem to have largely ignored it.

SP800-108 is kinda also in that bucket. But both SCP03 and Kerberos have found use for it, and neither with AES-XCBC.

I think I'd rather go with a (HMAC+CMAC) union and leave AES-XCBC as an outlier for IKE to implement. 

My 2c.

Reply:
In any case you should at least use some helper function to handle if the hmac is supported. You don't want to keep updating this code every time we add a new hash function.